docs/design-documents/hal/0002-pinmap-extension.md

```diff
@@ -0,0 +1,104 @@
+# HAL  PinMap
+
+<!-- toc -->
+
+- [Description](#Description)
+- [Motivation](#Motivation)
+- [Requirements](#Requirements)
+- [Design choices](#Design-choices)
+- [Questions](#Questions)
+
+<!-- tocstop -->
+
+## Description
+
+This update aims at increasing tests thoroughness by enabling them to check driver's reliability for each pin declared as supporting a peripheral.
+
+## Motivation
+
+At the time being, drivers are only tested on a single "default" peripheral. However, some target features the same peripheral through different IP for example the SPI may be provided on a single MCU by its USART, QSPI and SSP peripherals. To ensure that the driver implementation is valid for all these peripherals we want the CI to assess each pin in at least one pin configuration for each peripheral.
+
+## Requirements
+
+1. We want to list all pins for a function on a peripheral.
+2. We want to list all functions a pin can provide.
+3. We want to list all pins for a form-factor regardless of the function they can provide.
+4. We want a printable name for each pin and a method to get that name.
+5. We want a list the reserved pins that cannot be tested.
+6. We want to select a duplicate-free set a pin from a set of PinMap.
+
+Any of the list mentioned above may be empty if none match the criteria.
+
+## Design choices
+
+These requirements do not impact the current API and thus they can all be implemented as an extension to the existing API.   
+It is proposed to introduce the following elements :
+- `DEVICE_PIN_EXTENDED_SUPPORT` A `device_has` flag indicating the implementation of this extension.
+- The following part of the API provides the features needed for requirements 3, 4 and 5
+  ```c
+  enum pinmap_form_factor_ {
+      PINMAP_FORM_FACTOR_ARDUINO_ZERO,
+      PINMAP_FORM_FACTOR_ARDUINO_DUE,
+      PINMAP_FORM_FACTOR_NXP_FRDM,
+      PINMAP_FORM_FACTOR_ST_MORPHO,
+      PINMAP_FORM_FACTOR_MTB,
+  } pinmap_form_factor_t;
+
+  /// returns a static NC terminated array of pins.
+  const PinName *pinmap_form_factor_pins(pinmap_form_factor_t form_factor);
+  /// returns a static NC terminated array of pins.
+  const PinName *pinmap_restricted_pins();
+  /// returns a null (\0) terminated static character array representing the name of the pin.
+  const char *pinmap_pin_to_string(PinName pin);
+  ```
+- These generic function make use of the existing API to answer requirements 1, 2 and 6.
+  ```c
+  /// Returns the `n`'th pin supporting the `peripheral` in the `map`.
+  PinName pinmap_find_peripheral_pin(const PinMap *map, int peripheral, uint32_t n);
+
+  /// Returns a pin from ((whitelist ∩ map) − blacklist).
+  /// Typically, whitelist will be the form factor list of pins.
+  /// The blacklist may be a list of pins already visited and/or tied to another peripheral.
+  PinName pinmap_find_peripheral_pin(const PinMap *map, const PinName *whitelist,
```

Comment:
I hope that's a helper function that works on "sets" rather than something partners will need to do.

Reply:
Indeed these functions are not platform dependant.